src/rgw/rgw_op.cc

```diff
@@ -1571,6 +1571,92 @@ void RGWStatBucket::execute()
   }
 }
 
+int RGWGetBucketOplog::verify_permission()
+{
+  if (!verify_bucket_permission(s, RGW_PERM_READ))
+    return -EACCES;
+
+  return 0;
+}
+
+void RGWGetBucketOplog::execute()
+{
+  RGWAccessHandle h;
+  struct rgw_log_entry entry;
+  string st = s->info.args.get("start-time"),
+         et = s->info.args.get("end-time");
+  uint64_t epoch = 0;
+  uint64_t nsec = 0;
+  utime_t  ut_st, ut_et, ut_curr, ut_end;
+
+  http_ret = 0;
+  if (st.empty() || et.empty()) {
+    ldout(s->cct, 0) << "start-time and end-time must be non-null! " << st << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+
+  if (utime_t::parse_date(st, &epoch, &nsec) < 0) {
+    ldout(s->cct, 0) << "Error parsing date " << st << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+  ut_st = utime_t(epoch, nsec);
+
+  if (utime_t::parse_date(et, &epoch, &nsec) < 0) {
+    ldout(s->cct, 0) << "Error parsing date " << et << dendl;
+    http_ret = -EINVAL;
+    return;
+  }
+  ut_et = utime_t(epoch, nsec);
+  ldout(s->cct, 20) << "start-time:" << st << " end-time:" << et << dendl;
+
+  ut_curr = ut_st;
+  ut_end = ut_et + utime_t((time_t)3600, 0);
+  while (ut_curr <= ut_end) {
+    string oid;
+    char oid_date[128];
+
+    ut_curr.sprintf(oid_date, sizeof(oid_date));
+    oid_date[10] = '-';
+    oid_date[13] = 0;
+
+    oid = oid_date;
+    oid += "-";
+    oid += s->bucket.bucket_id;
+    oid += "-";
+    oid += s->bucket_name;
+    ldout(s->cct, 0) << "oid:" << oid << dendl;
+
+    int r = store->log_show_init(oid, &h);
+    if (r < 0) {
+        ldout(s->cct, 0) << "Error opening log " << oid << dendl;
+        http_ret = -EINVAL;
+        return;
+    }
+
+    do {
+      r = store->log_show_next(h, &entry);
+      if (r < 0 && r != -ENOENT) {
+        ldout(s->cct, 0) << "Error opening log " << oid << " r:" << r << dendl;
+        http_ret = -EINVAL;
+        return;
+      } else if (r == 0 || r == -ENOENT) {
+        break;
+      }
+
+      if (entry.time >= ut_st && entry.time <= ut_et) {
+        entries.push_back(entry);
```

Comment:
@sunfch since we keep all entries in memory, need to limit max entries we return (e.g., 1000). This is a good idea anyway, but better stream response out as we get it from the backend. Add a callback for writing out entry, and implement it (in addition to send_response()). We do it elsewhere in similar cases.